Moves Aborter class into separate package

chradek opened the pull request:
This PR builds on #2373 and moves the Aborter into its own package. Once this makes it into `master` we can update existing packages to use the Aborter from this package.

jeremymeng commented on sdk/core/aborter/tsconfig.json:
```diff
@@ -0,0 +1,39 @@
+    "sourceMap": true /* Generates corresponding '.map' file. */,
+    "outDir": "./dist-esm" /* Redirect output structure to the directory. */,
+    "stripInternal": true /* Do not emit declarations for code with @internal annotation*/,
+    "declarationDir": "./typings" /* Output directory for generated declaration files.*/,
```
I believe we now prefer `types` over `typings`.

chradek replied:
Updated to use `types` instead of `typings`.

jeremymeng commented on sdk/core/aborter/.prettierrc.json:
```diff
@@ -0,0 +1,6 @@
+{
```
We should use the existing `sdk/.prettierrc.json`

chradek replied:
Removed `.prettierrc.json` from the `core/aborter` package and now using the one located under `sdk`. Ran `check-format` and `format` scripts and committed those changes as well.

jeremymeng commented on sdk/core/aborter/package.json:
```diff
@@ -0,0 +1,105 @@
+    "check-format": "prettier --list-different --config .prettierrc.json \"src/**/*.ts\" \"test/**/*.ts\" \"*.{js,json}\"",
+    "clean": "rimraf dist dist-esm dist-test types temp browser/*.js* browser/*.zip statistics.html coverage coverage-browser .nyc_output *.tgz *.log test*.xml TEST*.xml",
+    "extract-api": "tsc -p . && api-extractor run --local",
+    "format": "prettier --write --config .prettierrc.json \"src/**/*.ts\" \"test/**/*.ts\" \"*.{js,json}\"",
```
`../../.prettierrc.json`

chradek replied:
Updated location in `check-format` and `format`. Thanks!

jeremymeng commented on sdk/core/aborter/src/aborter.ts:
```diff
@@ -0,0 +1,284 @@
+ *
+ * @example
+ * // Abort without timeout
+ * await blockBlobClient.upload(Aborter.none, buf, buf.length);
```
We might need to update examples as `blockBlobClient` is storage specific.

chradek replied:
That's fair, I think I should be able to adapt the examples in the readme to work for this documentation as well.

chradek replied:
Updated to be similar to the examples in the README.

jeremymeng commented:
Please check with @bterlson about the need to invetigate doc impact in the original PR.

chradek commented:
@bterlson 
Do you have any concerns on the documentation impact of this change? This PR does not include updating existing packages (e.g. storage-*) to consume this package.

bterlson commented:
As nothing has been updated to use the dependency yet, the documentation impact seems zero. But once we update a project's dependencies, we will want to check that the docs render/link appropriately. The docs for this package may need to be onboarded, the docs teams channel may be able to help with that process.

chradek commented:
@bterlson 
I talked to @ramya-rao-a and she mentioned the docs are published after we publish the package to NPM. I was able to verify that I could build the documentation for this package locally.

chradek commented:
@jeremymeng 
Now that CI issues are resolved, this should be ready for a final review. I've addressed your feedback, and made one additional change to rename the folder containing this package from `sdk/core/aborter` to `sdk/core/core-aborter`. This was done due to the feedback received for `core-http`.

jeremymeng commented on sdk/core/core-aborter/api-extractor.json:
```diff
@@ -0,0 +1,30 @@
+  },
+  "dtsRollup": {
+    "enabled": true,
+    "publicTrimmedFilePath": "./types/aborter.d.ts"
```
please add a line to disable generation of untrimmed dts rollup. otherwise it is generated by default into `dist/` and thus gets packaged.

```js
   "untrimmedFilePath": "",
```

See similar changes in https://github.com/Azure/azure-sdk-for-js/pull/3461/files
